File: ts/src/glob.ts
/**
 * Glob matching utilities for gitstore.
 *
 * Provides fnmatch-style matching with dotfile-aware semantics:
 * `*` and `?` do not match a leading `.` unless the pattern itself starts with `.`.
 */

/**
 * Match a single filename against a glob pattern segment.
 *
 * Supports `*`, `?`, `[seq]`, and `[!seq]`. Does not match across `/`.
 * A leading `.` in name is not matched by `*` or `?` unless the pattern
 * itself starts with `.` (Unix/rsync convention).
 */
export function globMatch(pattern: string, name: string): boolean {
  if (!pattern.startsWith('.') && name.startsWith('.')) {
    return false;
  }
  return fnmatch(pattern, name);
}

/**
 * fnmatch-style glob matching (single segment, no path separators).
 * Supports `*`, `?`, `[seq]`, `[!seq]`.
 */
function fnmatch(pattern: string, name: string): boolean {
  const regex = globToRegex(pattern);
  return regex.test(name);
}

function globToRegex(pattern: string): RegExp {
  let result = '^';
  let i = 0;
  while (i < pattern.length) {
    const ch = pattern[i];
    if (ch === '*') {
      result += '.*';
    } else if (ch === '?') {
      result += '.';
    } else if (ch === '[') {
      let j = i + 1;
      let negate = false;
      if (j < pattern.length && pattern[j] === '!') {
        negate = true;
        j++;
      }
      const end = pattern.indexOf(']', j);
      if (end < 0) {
        // No closing bracket — treat `[` as literal
        result += '\\[';
      } else {
        const chars = pattern.slice(j, end).replace(/[\\\]^]/g, '\\$&');
        result += negate ? `[^${chars}]` : `[${chars}]`;
        i = end;
      }
    } else if ('.+^${}()|\\'.includes(ch)) {
      result += '\\' + ch;
    } else {
      result += ch;
    }
    i++;
  }
  result += '$';
  return new RegExp(result);
}

File: cpp/src/glob.cpp
#include "internal.h"

namespace vost {
namespace glob {

/// Match a single pattern segment against a name.
/// Supports `*` (any sequence, not matching leading `.`),
///          `?` (any single char, not matching leading `.`),
///          `[...]` character classes.
bool fnmatch(const std::string& pattern, const std::string& name) {
    size_t pi = 0, ni = 0;
    size_t plen = pattern.size(), nlen = name.size();

    while (pi < plen && ni < nlen) {
        char pc = pattern[pi];
        if (pc == '*') {
            // Skip consecutive stars
            while (pi < plen && pattern[pi] == '*') ++pi;
            if (pi == plen) return true; // trailing * matches rest
            // Try matching rest of pattern at each position
            for (size_t k = ni; k <= nlen; ++k) {
                std::string rest_name = name.substr(k);
                std::string rest_pat  = pattern.substr(pi);
                if (fnmatch(rest_pat, rest_name)) return true;
            }
            return false;
        } else if (pc == '?') {
            ++pi; ++ni;
        } else if (pc == '[') {
            // Character class
            ++pi;
            bool negate = (pi < plen && pattern[pi] == '!');
            if (negate) ++pi;
            bool matched = false;
            char ch = name[ni];
            while (pi < plen && pattern[pi] != ']') {
                if (pi + 2 < plen && pattern[pi + 1] == '-') {
                    if (ch >= pattern[pi] && ch <= pattern[pi + 2])
                        matched = true;
                    pi += 3;
                } else {
                    if (ch == pattern[pi]) matched = true;
                    ++pi;
                }
            }
            if (pi < plen) ++pi; // skip ']'
            if (matched == negate) return false;
            ++ni;
        } else {
            if (pc != name[ni]) return false;
            ++pi; ++ni;
        }
    }

    // Consume trailing stars
    while (pi < plen && pattern[pi] == '*') ++pi;

    return pi == plen && ni == nlen;
}

/// Match a glob pattern against a string.
/// Leading dot in name requires explicit dot in pattern (unless pattern is `**`).
bool glob_match(const std::string& pattern, const std::string& name) {
    // For * and ? patterns, don't match names starting with '.'
    if (!name.empty() && name[0] == '.' &&
        !pattern.empty() && pattern[0] != '.') {
        // Exception: if pattern is just ** (for recursive glob)
        if (pattern != "**") return false;
    }
    return fnmatch(pattern, name);
}

} // namespace glob
} // namespace vost
